src/main/scala/uk/gov/nationalarchives/consignmentexport/Main.scala

```diff
@@ -21,39 +23,59 @@ object Main extends CommandIOApp("tdr-consignment-export", "Exports tdr files in
+        val exportFailedErrorMessage = s"Export for consignment $consignmentId failed"
+        val stepFunction = StepFunction(StepFunctionUtils(sfnAsyncClient))
+        val noTaskToken = !taskToken.isDefined
+
+        val exitCode = for {
+          config <- config()
+          rootLocation = config.efs.rootLocation
+          exportId = UUID.randomUUID
+          basePath = s"$rootLocation/$exportId"
+          tarPath = s"$basePath/$consignmentId.tar.gz"
+          bashCommands = BashCommands()
+          graphQlApi = GraphQlApi(config.api.url, config.auth.url)
+          keycloakClient = KeycloakClient(config)
+          s3Files = S3Files(S3Utils(s3Async))
+          bagit = Bagit()
+          validator = Validator(consignmentId)
+          //Export datetime generated as value needed in bag metadata and DB table
+          //Cannot use the value from DB table in bag metadata, as bag metadata created before bagging
+          //and cannot update DB until bag creation successfully completed
+          exportDatetime = ZonedDateTime.now(ZoneOffset.UTC)
+          consignmentResult <- graphQlApi.getConsignmentMetadata(config, consignmentId)
+          consignmentData <- IO.fromEither(validator.validateConsignmentResult(consignmentResult))
+          _ <- IO.fromEither(validator.validateConsignmentHasFiles(consignmentData))
+          bagMetadata <- BagMetadata(keycloakClient).generateMetadata(consignmentId, consignmentData, exportDatetime)
+          validatedFileMetadata <- IO.fromEither(validator.extractFileMetadata(consignmentData.files))
+          _ <- s3Files.downloadFiles(validatedFileMetadata, config.s3.cleanBucket, consignmentId, basePath)
+          bag <- bagit.createBag(consignmentId, basePath, bagMetadata)
+          fileMetadataCsv <- BagAdditionalFiles(bag.getRootDir).createFileMetadataCsv(validatedFileMetadata)
+          checksums <- ChecksumCalculator().calculateChecksums(fileMetadataCsv)
+          _ <- bagit.writeTagManifestRows(bag, checksums)
+          // The owner and group in the below command have no effect on the file permissions. It just makes tar idempotent
+          _ <- bashCommands.runCommand(s"tar --sort=name --owner=root:0 --group=root:0 --mtime ${java.time.LocalDate.now.toString} -C $basePath -c ./$consignmentId | gzip -n > $tarPath")
+          _ <- bashCommands.runCommand(s"sha256sum $tarPath > $tarPath.sha256")
+          _ <- s3Files.uploadFiles(config.s3.outputBucket, consignmentId, tarPath)
+          _ <- graphQlApi.updateExportLocation(config, consignmentId, s"s3://${config.s3.outputBucket}/$consignmentId.tar.gz", exportDatetime)
+          _ <- stepFunction.publishSuccess(taskToken, ExportOutput(consignmentData.userid))
+        } yield ExitCode.Success
+
+        exitCode.flatMap(ec => {
+          IO(ec)
+        }).recoverWith({
+          case rte: RuntimeException =>
+            //Continue to throw exception until taskToken required
+            if(noTaskToken) throw rte
```

Comment:
Is a task token required? If so, should we do some validation at the start of the job and fail it without running the export?

Reply:
As an initial step it the task token needs to be optional to ensure the export process is not interrupted with these changes. The deployment steps need to be:
1. Add the taskToken handling to the export process with the it as optional
2. Update to the Step Function to pass the task token to the ECS task and wait for it to return before continuing
3. Update the export process to make the taskToken as required

If I deploy the Step Function changes first, then the task token will never be returned by the ECS task and will hang forever. If I deploy the export changes first and have the task token as required, then the publish to the Step functions will error, as there is no valid token to return. It therefore needs the staged deployment. If that makes sense.

Reply:
Yeah, it makes sense that you'd need a staged deployment for that. Won't rethrowing the exception still cause the overall export to fail, though?

Reply:
It will, but that is the current behaviour of the export. Will remove this re-throw once the Step Function is updated to send the task token.

Reply:
Using the Step Function client method publish `sendTaskFailure` triggers the `Notify Failure` path of the Step Function. This logs the error in CloudWatch as a "Task Failed". 

However, the Step Function as a whole still succeeds. To fail the whole Step Function and log this, would need to add a "Fail State" step in the Step Function, and have the `Notify Failure` step call this. I think this will be relevant for https://national-archives.atlassian.net/browse/TDR-870. I had a bit of a fiddle with this. Looks like it might be a bit tricky to log any additional information in the Fail State step. Though pertinent information would be logged by the `Notify Failure` step. There are some examples in the Sandbox account.